File: LeetCode-Java/Combination Sum.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        int[] record = new int[candidates.length];
        for(int i = 0;i < record.length;i ++)
        {
            record[i] = 0;
        }
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        combinationSumPart(candidates,record,0, target,result);
        return result;
    }
        void combinationSumPart(int[] candidates,int[] record, int index, int target, List<List<Integer>> result)
        {
            if(index >= record.length) return;
            while(target >= 0)
            {
                if(target == 0)
                {
                    List<Integer> tempResult = new ArrayList<Integer>();
                    for(int i = 0;i < record.length;i ++)
                    {
                        for(int j = 0;j < record[i];j ++)
                        {
                            tempResult.add(candidates[i]);
                        }
                    }
                    result.add(tempResult);
                    break;
                }
                else
                {
                    combinationSumPart(candidates,record,index + 1,target,result);
                    record[index] = record[index] + 1;
                    target -= candidates[index];
                }
            }
            record[index] = 0;
            return;
        }
}

File: Leetcode-cpp/combination_sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        map<int, int> record;
        vector<vector<int>> result;
        vector<int> tempResult;
        search(result, candidates, tempResult, 0, target);
        return result;
    }
    void search(vector<vector<int>>& result, vector<int>& candidates, vector<int>& tempResult, int index, int target) {
        if (target == 0) {
            vector<int> currentResult(tempResult.begin(), tempResult.end());
            result.push_back(currentResult);
        } else if (target > 0 && index < candidates.size()) {
            int val = candidates[index];
            int pre_size = tempResult.size();
            while (target >= 0) {
                search(result, candidates, tempResult, index + 1, target);
                tempResult.push_back(val);
                target -= val;
            }
            int after_size = tempResult.size();
            tempResult.erase(tempResult.begin() + pre_size, tempResult.begin() + after_size);
        }
    }
};